src/config/crdt/types/datacenter_map.rs

```diff
@@ -0,0 +1,437 @@
+use crate::{
+    config::{
+        Datacenter, IcaoCode,
+        crdt::collections::{self, Cmrdt, ContentEqual, map::Op as MapOp},
+    },
+    generated::envoy::service::discovery::v3::Resource as XdsResource,
+    net::{NodeAddress, NodeIp, UnknownIp},
+};
+use std::{collections::HashSet, sync::Arc};
+
+pub type DatacenterMapCollection =
+    collections::Map<IcaoCode, collections::Orswot<NodeAddress, NodeIp>, NodeIp>;
+pub type InnerDatacenterMapOp = MapOp<IcaoCode, collections::Orswot<NodeAddress, NodeIp>, NodeIp>;
+
+impl typemap_rev::TypeMapKey for XdsDatacenterMap {
+    type Value = Arc<XdsDatacenterMap>;
+}
+
+impl crate::config::DynamicConfig {
+    #[inline]
+    pub fn xds_datacenters(&self) -> Option<&Arc<XdsDatacenterMap>> {
+        self.typemap.get::<XdsDatacenterMap>()
+    }
+}
+
+#[derive(Debug)]
+pub struct XdsDatacenterMap {
```

Comment:
nit: Can we remove `Xds` from the name? the information isn't xds specific, it could come from other providers (e.g. config file from fs).